README.md
# CFL Converter

[![Build Status](https://travis-ci.com/dhkatz/cfl-converter.svg?branch=master)](https://travis-ci.com/dhkatz/cfl-converter) [![npm version](https://badge.fury.io/js/cfl-converter.svg)](https://badge.fury.io/js/cfl-converter)

Convert from IMVU's CFL format to a file usable by the client create mode (.chkn).

In reality, a .chkn file is just a zip archive and can be opened as such.

Inspired by https://github.com/Toyz/LibCFL

## Usage

### Library

This package can also be integrated into your own program.

You can `import { convert } from 'cfl-converter';` and call the function with an `ArrayBufferLike` (such as a Buffer) containing the contents of
a CFL file (and any supported options in the second parameter).

A `decompress(data: ArrayBufferLike)` function is also provided if you would like to manually decompress the LZMA portions of the CFL.

### Command Line Tool

Convert .cfl files to .chkn files using this easy command-line tool!

Product IDs can also be supplied and the program will attempt to download them and convert them.

```
Usage: cfl-converter [options]

Options:
  -V, --version         output the version number
  -I, --input <files>   .CFL file(s) to convert.
  -P, --products <ids>  List of Product IDs to retrieve and convert.
  -h, --help            output usage information

Examples:
  $ cfl-converter --input product.cfl,chair.cfl
  $ cfl-converter --products 1243456,654321
```

## Testing

Testing is currently being written, although the program itself is quite simple.

## Details

The signifiance of this is, that due to what I can only call a CDN exploit, you can download the CFL for ANY product in the shop!

This can be accomplished by running the product ID through the url `http://userimages-akm.imvu.com/productdata/{productID}/1` where `{productID}` is the ID of the product you with to retrieve. 

The number at the end seems to be the revision number, so you may want to check if numbers higher than 1 work as well and get the highest revision you can.

After, simply save the downloaded file with a .cfl extension and use this tool to convert it to a CHKN file.

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

## Links

 * [Travis-ci](https://travis-ci.com/dhkatz/cfl-converter) 


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2020-12-27
### Fixed
- Fixed corrupted textures from downloaded CFLs.
### Changed
- Updated NodeJS dependencies.

## [0.2.2] - 2019-04-22
### Added
- Added bin option to allow usage from cli easily

## [0.2.1] - 2019-04-22
### Fixed
- Fixed bug with convert options throwing error if options are not passed

## [0.2.0] - 2019-04-22
### Changed
- Refactored much of the code to split into separate modules
- Changed dependencies in converter program to be optional

## [0.1.1] - 2019-04-20
### Changed
- Changed the program to suppor the new ReadBytes() method from csharp-binary-stream

## [0.1.0] - 2019-04-20
### Added
- Added CHANGELOG.md

[Unreleased]: https://github.com/dhkatz/cfl-converter/compare/1.0.0...HEAD
[1.0.0]: https://github.com/dhkatz/cfl-converter/compare/0.2.2...1.0.0
[0.2.2]: https://github.com/dhkatz/cfl-converter/compare/0.2.1...0.2.2
[0.2.1]: https://github.com/dhkatz/cfl-converter/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/dhkatz/cfl-converter/compare/0.1.1...0.2.0
[0.1.1]: https://github.com/dhkatz/cfl-converter/compare/0.1.0...0.1.1
[0.1.0]: https://github.com/dhkatz/cfl-converter/compare/0.1.0...0.1.1


gulpfile.ts
import del from 'del';

import sourcemaps from 'gulp-sourcemaps';
import typescript from 'gulp-typescript';
import merge from 'merge2';
import path from 'path';

import { dest, series, src, watch } from 'gulp';
import { CompileStream } from 'gulp-typescript';

const project = typescript.createProject('tsconfig.json', { declarationFiles: true });
const destination = 'build/';

function build() {
  del.sync([`${destination}**/*.*`]);

  const compiled: CompileStream = src(['src/**/*.ts'])
    .pipe(sourcemaps.init())
    .pipe(project());

  return merge([
    compiled.js
    .pipe(sourcemaps.write({ sourceRoot: (file) => path.relative(path.join(file.cwd, file.path), file.base) }))
    .pipe(dest(destination)),
    compiled.dts
      .pipe(dest(destination)),
  ]);
}

function update() {
  watch('src/**/*.ts', series(build));
}

exports.build = series(build);
exports.watch = series(build, update);
exports.default = series(build);


src/index.ts
import { BinaryReader, BinaryWriter, Encoding } from 'csharp-binary-stream';
import { buffer as toBuffer } from 'get-stream';

import JSZip from 'jszip';
import lzma from 'lzma-purejs';

/*
  Extra interfaces have been defined in case I decide to expand this package.
  I had plans to create a web interface to mimic the CFL Viewer that Toyz made,
  but I don't really see the point in wasting the effort when Toyz' program exists.
*/

export interface CFLEntry {
  compression: number;
  offset: number;
  length: number;
  size: number;

  name: string;
  hash: string | null;

  contents: ArrayBufferLike;
}

export interface CFLOptions {
  log?: boolean;
}

/**
 * Convert a CFL file buffer to a ZIP file buffer.
 * @param buffer A `Buffer` structure created from a CFL file.
 * @param options
 */
export async function convert(buffer: ArrayBufferLike, options: CFLOptions = {}): Promise<Buffer> {
  const reader = new BinaryReader(buffer);

  // CFL Header
  // This information is exclusive to the CFL format

  const hashed = reader.readChars(4, Encoding.Utf8) === 'DFL3';

  reader.position = reader.readUnsignedInt();

  const format = reader.readInt();

  if (options.log) {
    console.log('Decoded CFL header.');
  }

  // Decompress the LZMA section of the CFL file
  const content = Uint8Array.from(reader.readBytes(reader.readInt()));
  const data = new BinaryReader(format === 4 ? decompress(content) : content);

  if (options.log) {
    console.log('Decompressed CFL content.');
  }

  // Parse the CFL entries
  const entries: CFLEntry[] = [];

  let position = 0;
  while (data.length > position) {
    // CFL entry header data
    const size = data.readInt();
    const offset = data.readInt();
    const compression = data.readInt();
    const name = data.readChars(data.readShort(), Encoding.Utf8);

    reader.position = offset;

    // Decompress the LZMA section of the entry
    const entry = Uint8Array.from(reader.readBytes(reader.readUnsignedInt()));
    const contents = compression === 4 ? decompress(entry) : entry;

    // Calculate hash and entry size
    const hash: string | null = hashed ? data.readChars(data.readInt(), Encoding.Utf8) : null;
    const length = 14 + name.length + (hash !== null && hash.match(/^\s*$/) === null ? 4 + hash.length : 0);

    entries.push({
      compression,
      contents,
      hash,
      length,
      name,
      offset,
      size,
    });

    position += length;
  }

  if (options.log) {
    console.log(`Parsed ${entries.length} CFL ${entries.length !== 1 ? 'entries' : 'entry'}.`);
  }

  const zip = new JSZip();

  entries.forEach((entry: CFLEntry) => {
    zip.file(entry.name, entry.contents);
  });

  return toBuffer(zip.generateNodeStream({ type: 'nodebuffer', streamFiles: true }));
}

/**
 * Decompress an LZMA buffer.
 * @param data LZMA compressed data.
 */
export function decompress(data: ArrayBufferLike): ArrayBufferLike {
  const reader = new BinaryReader(data);
  const writer = new BinaryWriter();
  // Read the first 5 bytes, they are the properties
  // Set size to -1 to just read to EOS
  lzma.decompress(reader.readBytes(5), reader, writer, -1);
  return writer.toUint8Array();
}


src/cfl-converter.ts
#!/usr/bin/env node
import fs from 'fs';

import { convert } from '.';

import program from "commander";
import download from "download";

async function main() {
  program
    .version('1.0.0')
    .option('-I, --input <files>', '.CFL file(s) to convert.', (value: string): string[] => value.split(','), [])
    .option('-P, --products <ids>', 'List of Product IDs to retrieve and convert.', (value: string) => value.split(','), [])
    .option('-Q, --quiet', 'Stop the program from logging progress.', Boolean, false)
    .option('-D, --dry-run', 'Skip outputting to a file and just attempt to convert.', Boolean, false)
    .option('--ignore-missing', 'Ignore IDs that point to non-existent products.', Boolean, false)
    .option('--ignore-extension', 'Attempt to process all files regardless of extension.', Boolean, false);

  program.on('--help', () => {
    console.log('');
    console.log('Examples:');
    console.log('  $ cfl-converter --input product.cfl,chair.cfl');
    console.log('  $ cfl-converter --products 1243456,654321');
  });

  program.parse(process.argv);

  async function downloadProduct(id: number | string): Promise<Buffer | null> {
    let i = 100;
    for (; i > 0; i--) {
      try {
        const file = await download(`http://userimages-akm.imvu.com/productdata/${typeof id === 'number' ? id : id.trim()}/${i}`);

        return Buffer.from(file);
      } catch {

      }
    }

    return null;
  }

  const options = {
    input: program.input as string[],
    log: !program.quiet,
    products: program.products as string[],
  };

  if (options.products.length > 0 && options.log) {
    console.log(`Downloading ${options.products.length} CFL files...\n`);
  }

  const products = await Promise.all(options.products.map(downloadProduct));

  if (!program.ignoreMissing) {
    products.forEach((value) => {
      if (value == null) {
        console.error('One or more CFLs failed to download! (Use the --ignore-missing flag to override)');
      }
    });
  }

  if (!program.ignoreExtension) {
    options.input.forEach((value) => {
      if (!value.toLocaleLowerCase().endsWith('.cfl')) {
        console.error('This program only supports files with the .CFL extension! (Use the --ignore-extension flag to override)');
        process.exit(1);
      }
    });
  }

  const inputs = [...options.input.map((input) => fs.readFileSync(input.trim())), ...products];
  const names = [...options.input.map((input) => input.toLocaleLowerCase().endsWith('.cfl') ? input.slice(0, -4) : input), ...options.products];

  if (options.log) {
    console.log(`Beginning processing on ${inputs.length} file(s). . .\n`);
  }

  await Promise.all(inputs.map(async (input, index) => {
    if (input == null) {
      return;
    }

    if (options.log) {
      console.log(`Processing file ${names[index]}. . .`);
    }

    try {
      const file = await convert(input, { log: options.log });

      if (!program.dryRun) {
        fs.writeFileSync(`${names[index]}.chkn`, file);
        if (options.log) {
          console.log(`Wrote to ${names[index]}.chkn.\n`);
        }
      } else if (options.log) {
        console.log(`Successfully converted ${names[index]}.\n`);
      }
    } catch {
      console.log(`Failed processing ${names[index]}!\n`);
    }
  }));

  if (options.log) {
    console.log(`Finished processing ${inputs.length} file(s)!`);
  }
}

main();
